### ISSUE TITLE
low map, cant train a good result in my  datasets

### ISSUE THREAD

**Issue by angryhen** [zylo117/Yet-Another-EfficientDet-Pytorch#203]
I trained my datasets with 2 classes. First, I trained it withhead only, lr=1e-3, and then I trained the whole model, lr=1e-5. 
Val's class_loss is about 0.2, and reg_loss is 0.03, but my map results are as follows
```
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.019
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.040
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.016
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.011
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.044
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.024
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.076
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.131
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.089
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.156
```
please tell me how to do

**Comment by kiddliao** [comment]
update code
author have changed loss.py

**Comment by angryhen** [comment]
> update code
> author have changed loss.py

I git clone this project in 4.23, It's the latest

**Comment by zylo117** [comment]
can you share loss graph?
I think 1e-3 is too big for simple dataset like that.

**Comment by angryhen** [comment]
> 可以分享损失图吗？
> 我认为1e-3对于像这样的简单数据集来说太大了。
> 很容易看出损耗图是否过拟合

hi , 
I've tried it several times. Here are my actions and results
```
1. python train.py -c 0 -p custom--lr 1e-3 --batch_size 32 --load_weights weights/efficientdet-d0.pth  --num_epochs 200 --head_only True
Val. Epoch: 110/200. Classification loss: 0.59728. Regression loss: 0.05891. Total loss: 0.65619

 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.001
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.008
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.041
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.033
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.046

2. python train.py -c 0 -p custom --lr 1e-5 --batch_size 16 --load_weights logs/goods/efficientdet-d0_107_3000.pth --num_epochs 1000
Val. Epoch: 136/1000. Classification loss: 0.20790. Regression loss: 0.04673. Total loss: 0.25463

 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.001
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.005
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.001
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.001
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.004
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.016
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.041
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.037

![image](https://user-images.githubusercontent.com/24446065/80280441-9ff77580-8736-11ea-84eb-abbedd5f3106.png)
![image](https://user-images.githubusercontent.com/24446065/80280460-b1408200-8736-11ea-9514-f7821812705e.png)
![Uploading image.png…]()

**Comment by zylo117** [comment]
it's overfitting.

**Comment by angryhen** [comment]
I've been training for a short time so  can't believe it's over fitting. 
Do you have any suggestions？

**Comment by zylo117** [comment]
reduce lr

**Comment by angryhen** [comment]
> 减少lr

but the LR dropped to 1e-8 at the end of training

**Comment by zylo117** [comment]
But it had been overfitting from the very begining.

**Comment by YanqingWu** [comment]
> > 减少lr
> 
> but the LR dropped to 1e-8 at the end of training

have you fixed your problem?

**Comment by angryhen** [comment]
> > > 减少lr
> > 
> > 
> > 但是在训练结束时LR降至1e-8
> 
> 您解决了问题吗？

I think it's hard to train small object, but I  got a normal result in normal anchor size datasets

**Comment by angryhen** [comment]
> > > 减少lr
> > 
> > 
> > but the LR dropped to 1e-8 at the end of training
> 
> have you fixed your problem?

I think it's hard to train small object, but I  got a normal result in normal anchor size datasets

**Comment by Cli98** [comment]
> > > > 减少lr
> > > 
> > > 
> > > but the LR dropped to 1e-8 at the end of training
> > 
> > 
> > have you fixed your problem?
> 
> I think it's hard to train small object, but I got a normal result in normal anchor size datasets

In case of possible need to modify your anchor, please try this repo:
https://github.com/Cli98/anchor_computation_tool

**Comment by czzbb** [comment]
> it's overfitting. totally. from the very beginning

could u tell me how to identify whether overfitting please?

**Comment by czzbb** [comment]
> it's overfitting. totally. from the very beginn

Could you tell me how to determine whether overfitting please? cuz it seems well for me

**Comment by zylo117** [comment]
@czzbb  there is no certain standard for overfitting.

**Comment by czzbb** [comment]
I want to know why this curve means overfitting



---Original---

**Comment by iamastar88** [comment]
I also want to know why it means overfitting? In my understanding, the validation loss doesn't ascent, and obviously, the loss of validation set in the graph keeps decreasing, slightly higher than the loss of training set, that's a well process.

**Comment by heitorrapela** [comment]
@czzbb, your training loss is still decreasing while your validation loss is still on a plateau. It means that your model adapts the weights too much for the training set while not generalising well because it is not improving for unseen data. A good strategy is to stop the training based on the validation loss.
